src/network/clientpackethandler.cpp

```diff
@@ -1559,8 +1561,9 @@ void Client::handleCommand_MediaPush(NetworkPacket *pkt)
 	m_media_pushed_files.insert(filename);
 
 	// create a downloader for this file
-	auto downloader = new SingleMediaDownloader(cached);
-	m_pending_media_downloads.emplace_back(token, downloader);
+	std::unique_ptr<SingleMediaDownloader> downloader{
+		new SingleMediaDownloader(cached) };
+	m_pending_media_downloads.emplace_back(token, std::move(downloader));
```

Comment:
You can't move the downloader at this point, it is used on the lines below.

`m_pending_media_downloads.emplace_back(token, std::unique_ptr<T>(downloader));` might work here (leaving the declaration as-is)

Reply:
Would it work to use a shared pointer for this? Isn't it questionable that we go to the trouble of using a unique pointer and we hold on to the underlying raw pointer? That seems like a pretty high risk to me, especially since the unique pointer believes it is responsible for the deallocation of the resource it protects.

Reply:
It is risky, in theory. Nothing happens to the entry before the function returns and this is what the function already did before (with the implicit cast).

Reply:
I've done the extra work to use a shared pointer and to check that it won't do anything unexpected in other parts of the code. `std::make_shared` is available in C++11 so this is not a problem, and I've used direct initialization, in the hopes that I may find favor.